File: delayedqueue-jvm/src/main/kotlin/org/funfix/delayedqueue/jvm/JdbcConnectionConfig.kt
/*
 * Copyright 2026 Alexandru Nedelcu
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.funfix.delayedqueue.jvm

/**
 * Represents the configuration for a JDBC connection.
 *
 * @property url the JDBC connection URL
 * @property driver the JDBC driver to use
 * @property username optional username for authentication
 * @property password optional password for authentication
 * @property pool optional connection pool configuration
 */
@JvmRecord
public data class JdbcConnectionConfig
@JvmOverloads
constructor(
    val url: String,
    val driver: JdbcDriver,
    val username: String? = null,
    val password: String? = null,
    val pool: JdbcDatabasePoolConfig? = null,
)

File: delayedqueue-scala/src/main/scala/org/funfix/delayedqueue/scala/JdbcConnectionConfig.scala
/*
 * Copyright 2026 Alexandru Nedelcu
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.funfix.delayedqueue.scala

import org.funfix.delayedqueue.jvm

/** Represents the configuration for a JDBC connection.
  *
  * @param url
  *   the JDBC connection URL
  * @param driver
  *   the JDBC driver to use
  * @param username
  *   optional username for authentication
  * @param password
  *   optional password for authentication
  * @param pool
  *   optional connection pool configuration
  */
final case class JdbcConnectionConfig(
  url: String,
  driver: JdbcDriver,
  username: Option[String] = None,
  password: Option[String] = None,
  pool: Option[JdbcDatabasePoolConfig] = None
) {

  /** Converts this Scala JdbcConnectionConfig to a JVM JdbcConnectionConfig. */
  @SuppressWarnings(Array("org.wartremover.warts.Null")) // Java interop
  def asJava: jvm.JdbcConnectionConfig =
    new jvm.JdbcConnectionConfig(
      url,
      driver.asJava,
      username.orNull,
      password.orNull,
      pool.map(_.asJava).orNull
    )
}

object JdbcConnectionConfig {
  /** Converts a JVM JdbcConnectionConfig to a Scala JdbcConnectionConfig. */
  def fromJava(javaConfig: jvm.JdbcConnectionConfig): JdbcConnectionConfig =
    JdbcConnectionConfig(
      url = javaConfig.url,
      driver = JdbcDriver.fromJava(javaConfig.driver),
      username = Option(javaConfig.username),
      password = Option(javaConfig.password),
      pool = Option(javaConfig.pool).map(JdbcDatabasePoolConfig.fromJava)
    )
}
